cassandra-analytics-core/src/main/java/org/apache/cassandra/spark/bulkwriter/coordinatedwrite/CassandraClusterInfoGroup.java

```diff
@@ -161,40 +160,22 @@ public void checkBulkWriterIsEnabledOrThrow()
     }
 
     /**
-     * @return the largest time skew retrieved from the target replicas
+     * @return the largest time skew retrieved from the target clusters
      */
     @Override
-    public TimeSkewResponse getTimeSkew(List<RingInstance> instances)
+    public TimeSkewResponse timeSkew(Range<BigInteger> range)
     {
         if (clusterInfos.size() == 1)
         {
-            return clusterInfos.get(0).getTimeSkew(instances);
+            return clusterInfos.get(0).timeSkew(range);
         }
 
-        Map<String, List<RingInstance>> instancesByClusterId = instances.stream().collect(Collectors.groupingBy(instance -> {
-            String clusterId = instance.clusterId();
-            Preconditions.checkState(clusterId != null,
-                                     "RingInstance must define its clusterId for coordinated write");
-            return clusterId;
-        }));
-        long localNow = System.currentTimeMillis();
-        long maxDiff = 0;
-        TimeSkewResponse largestSkew = null;
-        for (Map.Entry<String, List<RingInstance>> entry : instancesByClusterId.entrySet())
-        {
-            String clusterId = entry.getKey();
-            List<RingInstance> instancesOfCluster = entry.getValue();
-            ClusterInfo clusterInfo = cluster(clusterId);
-            Preconditions.checkState(clusterInfo != null, "ClusterInfo not found with clusterId: " + clusterId);
-            TimeSkewResponse response = clusterInfo.getTimeSkew(instancesOfCluster);
-            long d = Math.abs(response.currentTime - localNow);
-            if (Math.abs(response.currentTime - localNow) > maxDiff)
-            {
-                maxDiff = d;
-                largestSkew = response;
-            }
-        }
-        return largestSkew;
+        return clusterInfos.stream()
+                           .map(clusterInfo -> clusterInfo.timeSkew(range))
+                            // Find the timeSkew with the lowest remote currentTime, i.e. largest difference with the local current time.
```

Comment:
This isn't really right, as it would be possible for the current time on a remote host to be much _greater_ than the current time on the host running the task. We really need to do the whole "take the absolute value of the difference between the times" thing, not just look for the min currentTime.

I realize also that for multi-cluster scenarios, the other field in the TimeSkewResponse (`allowableSkewInMinutes`) can, in fact, be different (although in theory I'd doubt that it would be). We should really grab 2 things:
1) The time skew response with the maximum _difference_ between localNow and the remote time, where localNow is calculated each time you get a response back (per my previous suggestion)
2) The _minimum_ `allowableSkewInMinutes` value

And then craft the "combined" time skew response from those two things.

Reply:
Talked offline. We are going to refactor the method to do the validation, instead of exposing a response object.